### ISSUE TITLE
[Bug] Embedding, head and norm parameters get gradients but never update in the quantized per-layer path

### ISSUE THREAD

**Issue by gss10282023** [VITA-Group/Q-GaLore#12]
### Summary

In the `q_galore_adamw8bit_per_layer` path of `run_pretrain.py`, the embedding, output head and norm weights receive nonzero gradients but stay exactly at their initial values.

The script creates an optimizer and scheduler for each of these parameters, but attaches the update function as a plain attribute: `setattr(p, "backward_hook", optimizer_hook)`. The custom quantized backward calls that attribute explicitly. Ordinary autograd on a floating-point tensor does not, so the optimizers for those parameters are never called.

This happens with gradient accumulation 1, so it is separate from any accumulation-boundary issue.

### Environment

Commit [5bd2f1c5](https://github.com/VITA-Group/Q-GaLore/blob/5bd2f1c5a1aded31223d7f15c4604c4d56cfe1e6/run_pretrain.py), PyTorch 2.3.0+cu121, bitsandbytes 0.43.1, Transformers 4.41.0, RTX A6000.

### Steps to reproduce

Run `run_pretrain.py` with a two-layer LLaMA (hidden size 128, vocabulary 512), projection rank 32, four fixed sequences of length 32, the quantized per-layer optimizer, deterministic rounding, FP32, gradient accumulation 1, no gradient clipping. Do one optimizer update and count optimizer and scheduler steps per parameter.

| Parameter group | Count | Optimizer steps | Scheduler steps |
| --- | ---: | ---: | ---: |
| Quantized matrix weights | 14 | 1 each | 1 each |
| Floating embedding, head and norm weights | 7 | 0 each | 0 each |

All seven floating parameters have nonzero gradients. Their expected updates are about `0.005` in max-abs; their actual updates are zero.

After registering a real autograd hook for the floating parameters, the full result (model, quantization metadata, optimizer state, scheduler state, remaining gradients) matches a driver that runs backward and then steps every optimizer explicitly. Reproduced on a second A6000 host.

### Expected behavior

Every parameter that gets an optimizer should also get stepped.

### Proposed fix

Register an autograd hook for ordinary trainable parameters and keep the attribute for quantized parameters, whose custom backward calls it:

```python
if p.requires_grad:
    p.register_post_accumulate_grad_hook(optimizer_hook)
else:
    setattr(p, "backward_hook", optimizer_hook)
```

This is the change I tested inside the existing parameter loop. It does not address the separate behavior of stepping on every backward when gradient accumulation is greater than 1. The test is a small FP32 model, not a reproduction of the paper's 7B results or memory numbers.
